tmt/steps/report/reportportal.py

```diff
@@ -1,152 +1,221 @@
-                             compresslevel=1) as zipstream:
-            # XML file names are irrelevant to ReportPortal
-            with zipstream.open("tests.xml", "w") as entry:
-                entry.write(data.encode("utf-8"))
-        bytestream.seek(0)
-
-        # Send the report to the ReportPortal instance
+        assert self.step.plan.name is not None
+        launch_name = self.get("launch") or self.step.plan.name
+
+        version = "v1"
+        url = f"{server}/api/{version}/{project}"
+        headers = {"Authorization": "bearer " + token,
+                   "accept": "*/*",
+                   "Content-Type": "application/json"}
+
+        tmt_vars = self.get("tmt-vars")
+        attributes = [{'key': k, 'value': v[0]} for k, v in self.step.plan._fmf_context.items()]
+
+        # Communication with RP instance
         with tmt.utils.retry_session() as session:
-            url = f"{server}/api/v1/{project}/launch/import"
-            self.debug(f"Send the report to '{url}'.")
-            response = session.post(
-                url,
-                headers={
-                    "Authorization": "bearer " + token,
-                    "accept": "*/*",
-                    },
-                files={
-                    # The zip filename is used as the launch name in ReportPortal
-                    "file": (launch_name + ".zip", bytestream, "application/zip"),
-                    },
-                )
-
-        # Handle the response
-        try:
-            message = tmt.utils.yaml_to_dict(response.text).get("message")
-        except (tmt.utils.GeneralError, KeyError):
-            message = response.text
-        if not response.ok:
-            raise tmt.utils.ReportError(
-                f"Received non-ok status code from ReportPortal, response text is: {message}")
 
-        self.debug(f"Response code from the server: {response.status_code}")
-        self.debug(f"Message from the server: {message}")
-        self.info("report", "Successfully uploaded.", "yellow")
+            # Create a launch
+            self.info("launch", launch_name, color="cyan")
+            response = session.post(url=f"{url}/launch",
+                                    headers=headers,
+                                    json={"name": launch_name,
+                                          "description": self.step.plan.summary,
+                                          "attributes": attributes,
+                                          "startTime": timestamp()})
+            self.handle_response(response)
+            launch_uuid = yaml_to_dict(response.text).get("id")
+            assert launch_uuid is not None
+            self.info("uuid", launch_uuid, "yellow", shift=1)
+
+            # For each test
+            for result in self.step.plan.execute.results():
+                test = [test for test in self.step.plan.discover.tests()
+                        if test.serialnumber == result.serialnumber][0]
```

Comment:
Maybe we should provide a `Result.test` attribute to make this less hackish - results are often created from a particular `Test` instance, therefore saving the instance in the result should be trivial.

Could you add a TODO for me, to resolve this? I guess there's no point in blocking your PR on this, but I'd like to fix it & I forget easily.

Reply:
Will do! Sorry for silly questions :)